### DISCUSSION TITLE
How to assign value to AsyncLocal before test so it is correctly propagated to test method?

### DISCUSSION THREAD

**Discussion by Laxynium** [thomhurst/TUnit#1278] [Q&A] [answered]
I've tried to start a new activity `ActivitySource.StartActivity` before each of my integration tests using `Before(Test)` attribute and then use 
```csharp
UseTestServer(opt =>
{
    opt.PreserveExecutionContext = true;
})
```
so it would be propagated into controller and asp.net core would treat that activity as parent.
However, that approach does not work, I'm guessing because Before is run in its own Task that is parallel to test method Task.
Here is the sample that uses only AsyncLocal, since Activities are using it underneath.
```csharp
public class AsyncLocalTest
{
    private readonly AsyncLocal<string> _asyncLocalValue = new();

    [Before(Test)]
    public void Before()
    {
        _asyncLocalValue.Value = "123";
    }

    [After(Test)]
    public void After()
    {
        TestContext.Current?.GetDefaultLogger().LogInformation($"Message: {_asyncLocalValue.Value}");
    }

    [Test]
    public async Task Test_method()
    {
        await Assert.That(_asyncLocalValue.Value).IsEqualTo("123");
    }
}
```
I've tried to make Before and After methods synchronous, but that didn't work.
Is there currently a way to some how achieve that, except of starting that activity inside each test explicitly?

**Comment by thomhurst** [comment]
I'm not sure it's possible. AFAIK setting AsyncLocal only propagates from outer to inner / parent to child, and can't do the opposite.

Because a Before hook isn't a 'parent' of the main test (more of a sibling I guess) then the async local can't propagate to it?

**Reply by viceroypenguin** [reply]
@thomhurst that is a correct interpretation. That's bit us on the behind too, because we can't set the IHttpContextAccessor HttpContext in the before either, since it relies on AsyncLocal as well. 

Possible solutions I can think of:  

a) offer a sync version of "before" (since sync shares the same async context as the parent, it could qualify as the parent as long as the parent method vis-a-vis AsyncLocal, or 

b) add a method that can be called from the before that would run the test directly, and the infra could determine when before completes if the test wasn't run yet, and only run if before didn't call it.

**Reply by thomhurst** [reply, marked as answer]
Option a sounds cleaner and a smoother experience for the end user. Would be a bit of a refactor of code though, but I'll give it a go!

**Comment by Laxynium** [comment]
Thanks for quick response. I'm looking forward to the new hook method.

**Comment by thomhurst** [comment]
@Laxynium @viceroypenguin could you try v0.4.0 and setting AsyncLocal values in a `void` `Before(Test)` hook?

**Reply by Laxynium** [reply]
I've tried, and it's working fine now.

**Reply by thomhurst** [reply]
Brilliant, cheers!
